### ISSUE TITLE
jest-expo fails on react-native

### ISSUE THREAD

**Issue by afeld** [expo/expo#3133]
### Environment

```
  Expo CLI 2.6.14 environment info:
    System:
      OS: macOS 10.14.2
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 10.9.0 - ~/.nvm/versions/node/v10.9.0/bin/node
      npm: 6.5.0 - ~/.nvm/versions/node/v10.9.0/bin/npm
    IDEs:
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^32.0.0 => 32.0.0
      react: 16.5.0 => 16.5.0
      react-native: https://github.com/expo/react-native/archive/sdk-32.0.0.tar.gz => 0.57.1
    npmGlobalPackages:
      expo-cli: 2.6.14
```

### Steps to Reproduce

```sh
expo init -t blank --npm first-project
cd first-project
npm i --save-dev jest-expo
echo 'module.exports = { preset: "jest-expo" };' > jest.config.js
touch App.test.js
./node_modules/.bin/jest
```

### Expected Behavior

Tests would pass.

### Actual Behavior

```
 FAIL  ./App.test.js
  ● Test suite failed to run

    /Users/aidan/dev/first-project/node_modules/react-native/jest/mockComponent.js: Unexpected token (20:23)

      Jest encountered an unexpected token
      This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.
      By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".
      Here's what you can do:
       • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
       • If you need a custom transformation specify a "transform" option in your config.
       • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.
      You'll find more details and examples of these config options in the docs:
      https://jestjs.io/docs/en/configuration.html
      Details:
        18 |
        19 |   const Component = class extends SuperClass {
      > 20 |     static displayName = 'Component';
           |                        ^
        21 |
        22 |     render() {
        23 |       const name =

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.676s
Ran all test suites.
```

Seems like this _should_ be working, as [the `transformIgnorePatterns` are in place](https://github.com/expo/expo/blob/84abc25243e88d49f6060d7d35ae074ba2c0420b/packages/jest-expo/jest-preset.json#L24-L26). @dalcib ideas?

### Reproducible Demo

See above - should be able to easily recreate. Package versions:

```
$ npm list --depth=0
/Users/aidan/dev/first-project
├── babel-preset-expo@5.0.0
├── expo@32.0.0
├── jest-expo@32.0.0
├── react@16.5.0
└── react-native@0.57.1
```

**Comment by afeld** [comment]
_Realized this isn't a TypeScript-specific problem, so updated the issue._

**Comment by sjchmiela** [comment]
Most interestingly… I was able to reproduce it only once. 😅 Then, after `rm -rf`ing `node_modules`, reinstalling the dependencies it started to work. Then, to try to reproduce the bug, I created:
- project with dependencies installed with `yarn`, then `yarn add -D jest-expo` (and other steps) and it worked
- project with dependencies installed with `npm`, then `npm i --save-dev jest-expo` and it worked too
- project with dependencies intalled with `yarn`, then `npm i --save-dev jest-expo`, it wasn't working with `Module react-native/jest/hasteImpl.js in the haste.hasteImplModulePath option was not found.` error message, but then I ran `yarn` and it worked again.

Could you confirm that the instructions you've provided still produce an invalid project environment?

**Comment by afeld** [comment]
Oof, so: when I do the steps at the top using `npm`, it still fails the same way. _But_, when I do them using `yarn`...it works.

> after `rm -rf`ing `node_modules`, reinstalling the dependencies it started to work

Hmm, not for me, at least when using `npm`.

> I was able to reproduce it only once.

This is sounding like a problem of [NPM's non-determinism](https://npm.github.io/how-npm-works-docs/npm3/non-determinism.html) 😓 

I tried comparing the diffs of `npm list` for the projects created the two different ways, but there were so many `dedupe`s and other differences that I couldn't make sense of it. Not really sure what to do next, other than just "use Yarn".

**Comment by SimenB** [comment]
It only really works in yarn as well due to hoisting (so it's most probably broken using yarn pnp). I haven't dug into how Expo sets up Jest, but it's pretty brittle as is.

The fix is to configure `babel-jest` correctly for babel 7 (that is - adding it to the top level `devDependencies` and adding the bridge package for babel).

```sh-session
$ expo init -t blank --npm first-project
$ cd first-project
$ npm i --save-dev jest-expo babel-core@bridge babel-jest
$ echo 'module.exports = { preset: "jest-expo" };' > jest.config.js
$ touch App.test.js
$ ./node_modules/.bin/jest
```

Only difference from OP is the install command (`npm i --save-dev jest-expo babel-core@bridge babel-jest`)

Results in

```
$ /Users/simen/repos/first-project/node_modules/.bin/jest
 FAIL  ./App.test.js
  ● Test suite failed to run

    Your test suite must contain at least one test.

      at node_modules/jest-expo/node_modules/jest/node_modules/jest-cli/build/TestScheduler.js:256:22
```

---

The `bridge` module is not needed in Jest 24 (currently in alpha) as we've upgraded to Babel 7, but you should still install `babel-jest` explicitly so Jest loads _your_ version of `babel`

**Comment by SimenB** [comment]
I guess `jest-expo` could have a peer dependency on `babel-jest`, instead of a `dependency`: https://github.com/expo/expo/blob/faa67b2fe9c13e906991af8b110fe7090359433f/packages/jest-expo/package.json#L22-L23

Then yarn/npm would warn you that you have to install it. Same with the bridge package for Jest 23 (but again, that can be removed for Jest 24)

**Comment by zakuni** [comment]
Unfortunately, adding `babel-jest` wasn't enough for me in project already exists.

But adding transform option to jest.config.js, as shown in https://github.com/expo/expo/issues/2595#issuecomment-440966998 helped me.

```js
module.exports = {
  preset: 'jest-expo',
  transform: {
    '\\.js$': '<rootDir>/node_modules/react-native/jest/preprocessor.js',
  }
};
```
`babel-core@bridge` isn't necessary anymore.

**Comment by jdhorner** [comment]
I _still_ have this issue with the default "Tabs" app that's created in the command line, as soon as any of the components are changed from being `.js` to `.jsx`. For the life of me, I can't find help for this problem anywhere. I put [more details here](https://forums.expo.io/t/testing-jsx-files-with-jest/22643), but perhaps it should be a github issue instead?

**Comment by Jlexyc** [comment]
The same issue, trying to fix. Jest 24 :(

**Comment by omerdn1** [comment]
I tried both @SimenB and @zakuni solutions and they didn't work. any updates on this?

**Comment by EvanBacon** [comment]
Fixed in #6478
